Sync analyze cache split

thearthur opened the pull request:
Split the fixed daily sync+analyse into independently scheduled
* sync-schema (discover table structure)
* cache sample field values
* analyze (create fingerprints)
* classify (guess field special types)

TBD
- [ ] ~classify needs  locking~
- [ ] ~analyze needs  locking~
- [x] create different default schedules for each type
- [ ] infer-special-type is still being used in cache field values
- [ ] ~using h2 test db still broken in repl~
- [x] remove debug prints
- [x] improve field % email/json to return actual %'s vs. preserving the existing "all or nothing" interpretation. 
- [ ] ~add updated-at timestamps to field values to match the other models~
- [ ] ~add function for caching values for a single field~
- [x] unschedule sync tasks when a DB is deleted!
- [x] stop re-syncing the sample-db (unless we want to start doing this)
- [x] make sure new DB's sync, cache, analyse, and classify upon being added.
- [x] rename "sync" phase to sync-schema, and name the overall process "Sync"
- [x] rename metabase.task.sync-cache-analyze-classify-databases
- [x] tasks need to be properly scheduled on DB create
- [x] add emoji progress bars to classify

thearthur committed:
split analize into analize and classify

thearthur committed:
move speceial type inference to classify

make notify trigger re-analysis
We should decide if we want this to happen, and for now it
fixes some tests that make this assumption.

thearthur committed:
remove pecent url calculation from driver

This was a source of full table scans.
split guessing special type by name to it's own namespace

thearthur committed:
store field and table fingerprints in DB

thearthur committed:
WIP, GIU for scheduling analyze and classify

thearthur committed:
use per-database scheduleing for sync, cache, analyze, and classify

and add basic scheduling to the DB setup page

thearthur committed:
remove longer used task file

All the sync/cache/analyze/classify db tasks are now defined in a single
file so the other task files like sync_database.clj are no longer
required.

thearthur committed:
ugly UI for scheduling sync tasks.

This UI doen't yet match the spec, though it work.

salsakran commented on src/metabase/api/table.clj:
```diff
@@ -66,14 +71,18 @@
+          driver (->> id table/table-id->database-id driver/database-id->driver) ;; is this the easy way to do this?
           new-visibility (visible-state? (:visibility_type updated-table))
           old-visibility (visible-state? original-visibility-type)
           visibility-changed? (and (not= new-visibility
```
`visibility-changed?` should probably be renamed to `table-now-visible` or something similar.

thearthur replied:
fixed

thearthur commented:
the driver "analyse" function is no longer used in the analyse phase. it should be renamed

thearthur commented:
some protection against starting too many tasks at once should be added

thearthur commented:
the start times should be randomized a bit, to avoid thread starvation

thearthur commented:
inferring the special type by name is happening twice, once in sync and once in classify, this should be unthreaded so it can happen once.

salsakran commented on frontend/src/metabase/components/DatabaseDetailsForm.jsx:
```diff
@@ -256,14 +260,41 @@ export default class DatabaseDetailsForm extends Component {
             {
                 name: 'name',
-                'display-name': 'Name',
+                'display-name': 'NameFOOsss',
```
is this a typo?

thearthur replied:
yes, from fussing with build-hot failing

thearthur replied:
fixed

salsakran commented on src/metabase/api/table.clj:
```diff
@@ -66,14 +71,18 @@
         (log/debug (u/format-color 'green "Table visibility changed, resyncing %s -> %s : %s") original-visibility-type visibility_type visibility-changed?)
-        (sync-database/sync-table! updated-table))
+        (sync-database/sync-table! updated-table)
+        (cached-values/cache-table-data-shape! driver updated-table)
```
`cache-table-data-shape!`  should probably figure out the driver (if it needs it) from the `updated-table`

salsakran commented on src/metabase/sync_database/analyze.clj:
```diff
@@ -157,86 +59,99 @@
+                     :is_fk                   (isa? (:special_type field) :type/FK)
+                     :cardinality             (count (distinct values))
+                     :field_percent_urls      (percent-valid-urls values)
+                     :field_percent_json      (if (values-are-valid-json? values) 100 0)
```
this and `field_percent_email` should probably actually unwind these functions and compute the actual %, vs picking an extreme of 0 or 100

thearthur replied:
added to the todo list

salsakran commented on src/metabase/sync_database/analyze.clj:
```diff
@@ -157,86 +59,99 @@
+                          field-values-count))))))
+
+(defn field-fingerprint [driver table field]
+  (let [values      (->> (driver/field-values-lazy-seq driver field)
```
We should feed this into cache-values at this point. Otherwise, we're doing double the work of the current process for no real gain.

salsakran replied:
This btw is probably the highest priority change/fix in the current batch in terms of getting things merged.

thearthur replied:
These values are independent of field cached values. 
There are a bunch of rules governing when field values are gathered and then they are generated by the driver. This section is totally independent of this, unaware of the rules for caching values (privacy etc), and happens after the cached values are set. These values are never stored beyond what goes into the fingerprint so we don't need the privacy controls. 
I would like to preserve the independence of the sample method for guessing special type so we can change how we choose a sample set without worrying about other subsystems.

salsakran commented on src/metabase/sync_database/cached_values.clj:
```diff
@@ -0,0 +1,157 @@
+      (assoc field-stats :values distinct-values)
+      field-stats)))
+
+(defn make-analyze-table
```
Think these should be renamed.

thearthur replied:
yes, they are now used only for caching values. No point in preserving the name just because the iterative refactor flow landed it that way.

salsakran commented on src/metabase/sync_database/cached_values.clj:
```diff
@@ -0,0 +1,157 @@
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;;; call drivers, get field values, save them in application DB
+(defn cache-table-data-shape!
```
We should also have a function to cache the values for a single field. It would be really useful if it accepted a flag for a sampled vs full scan.

salsakran commented on src/metabase/driver.clj:
```diff
@@ -20,7 +20,7 @@
    This many is probably fine for inferring special types and what-not; we don't want
    to scan millions of values at any rate."
-  10000)
+  1000)
```
let's go back to 10k

thearthur replied:
fixed

salsakran commented on src/metabase/sync_database/classify.clj:
```diff
@@ -0,0 +1,204 @@
+
+(defn classify-tables!
+  "classify and save all previously saved fingerprints for tables in this database"
+  [driver {database-id :id, :as database}]
```
do we need a driver here?

thearthur replied:
nope, and another function falls out of the callgraph when it's removed. good catch!

salsakran commented on src/metabase/task/sync_cache_analyze_classify_databases.clj:
```diff
@@ -0,0 +1,101 @@
+(ns metabase.task.sync-cache-analyze-classify-databases
```
This should have a simple/shorter name. 
"Analyze Databases" or "Collect Metadata" perhaps.

thearthur committed:
cleaning up and renaming refactord functions

thearthur committed:
code cleanup and fix test-data

thearthur committed:
cleanup comments and logging

salsakran commented:
@camsaul or @thearthur -- I forget, where does the sync process get kicked off when a new DB is added? couldn't find it in the models or api database.clj files.

salsakran commented on src/metabase/sync_database/analyze.clj:
```diff
@@ -259,3 +175,11 @@
+  "analyze all the tables in one database"
+  [db]
+  (analyze-data-shape-for-tables! (->> db
+                                       :id
```
I might be missing something, but my read, this has resulted in some missing behaviour when adding a DB.

Right now, when you add a DB, the `process-sync-database-event` handler is triggered in https://github.com/metabase/metabase/blob/sync-analyze-cache-split/src/metabase/events/sync_database.clj#L22

this calls 'sync-database!' https://github.com/metabase/metabase/blob/sync-analyze-cache-split/src/metabase/sync_database.clj#L68

which ends up in 'sync-database-with-tracking'
https://github.com/metabase/metabase/blob/sync-analyze-cache-split/src/metabase/sync_database.clj#L39

This however, only does an "analysis" unless I am mistaken. 

This seems to break the current sequence for adding a database where 
1. the schema is sync'd
2. the analysis step does the equivalent of the new [cache-values, analyze, classify]
and instead, the value caching and classification only happens when the nightly process finds them.

This is a regression and  not a good outcome for users as new databases aren't really useable when added.

To rectify this I think
1. Classification is only useful when new information appears. Practically, this means a new fingerprint has been generated. Hence, it should be called from the analyze task.
2. When adding a new DB, we should schedule both the chained (analyze + classify) and cache-values tasks for immediate execution.

thearthur replied:
sounds like you found a bug, I'll make sure this works. I'll make sure this to works as it does on master now.

salsakran replied:
also, this lends me to believe that there shouldn't be a separate schedule for classification at this point. it should only run when triggered by a new fingerprint being available or when new classifiers are shipped (i.e. possibly a startup check if we ship new ones in the future).

salsakran commented on src/metabase/sync_database/cached_values.clj:
```diff
@@ -0,0 +1,136 @@
+  [{:keys [base_type] :as field}]
+  (or (field-values/field-should-have-field-values? field)
+      (and (not (isa? base_type :type/DateTime))
+           (not (isa? base_type :type/Collection))
```
Think we should probably add the below to this list. Can't see them ever being useful for filter autocompletes
:type/Float
:type/URL
:type/Email
:type/Description
:type/TextLike
:type/PK
:type/FK
 
thoughts @mazameli or @kdoh ?

salsakran replied:
(or more accurately, this list might update `should-have-field-values` at https://github.com/metabase/metabase/blob/sync-analyze-cache-split/src/metabase/models/field_values.clj#L30

salsakran replied:
this specific function only seems to be called from the mongo driver.

camsaul replied:
from a readability standpoint it really might just make sense to derive the types in question here from a one-off namespaced type and then use that for the checks. e.g. something like

```clojure
(derive ::skip-cardinality-testing :type/DateTime)
(derive ::skip-cardinality-testing :type/Collection)
...
;; Don't test base types that derive from ::skip-cardinality-testing
(not (isa? base_type ::skip-cardinality-testing))
  ...)
```

thearthur committed:
use differnt default schedules for each sync phase.

salsakran commented:
Question -- 

would it make sense to rename the current "sync" process to "sync-schema" or something similar, and name the entire set of sync, analyze, cache, etc processes "sync"? Or have some other name that captures the entire set of processes.

thearthur commented:
I do like the idea of renaming "sync"| to have a name specific to just this phase of the "sync" process.
keeping the name "Sync" for the overall process makes sense to me, and having the "sync" endpoint already exposed run the entire process start to finish perhaps?

thearthur committed:
rename the sync phase of sync/cache/analysis/classify to sync-schema

also rename the overall task to just 'sync'

thearthur committed:
fix eventing around sync

fully sync new databases
property schedule sync on newly created databases
unschedule a database when it's removed

thearthur committed:
don't schedule sync for the sample db

thearthur committed:
correct titles on progress bars in sync

camsaul commented on src/metabase/api/database.clj:
```diff
@@ -272,10 +274,16 @@
+                           :classify_schedule  classify_schedule
                            :description        description
                            :caveats            caveats
                            :points_of_interest points_of_interest)) ; TODO - this means one cannot unset the description. Does that matter?
```
If you hit `C-c SPC` in Emacs it will nicely indent the map for you

camsaul commented on src/metabase/cache_database.clj:
```diff
@@ -0,0 +1,77 @@
+            [metabase.sync-database.cached-values :as cached-values]
+            [toucan.db :as db]))
+
+(defonce ^:private currently-syncing-dbs (atom #{}))
```
what goes in this atom? Database IDs?

camsaul commented on src/metabase/cmd/load_from_h2.clj:
```diff
@@ -37,6 +37,8 @@
              [field :refer [Field]]
              [field-values :refer [FieldValues]]
+             [field-fingerprint :refer [FieldFingerprint]]
+             [table-fingerprint :refer [TableFingerprint]]
```
`:require` ordering 😉

camsaul commented on src/metabase/cmd/load_from_h2.clj:
```diff
@@ -100,6 +102,8 @@
    CollectionRevision
    DashboardFavorite
+   FieldFingerprint
+   TableFingerprint
```
keeping this list alphabetized will make things a little easier, help us avoid duplicates, etc.

thearthur replied:
The comment below indicates that order may be important, or is it only DataMigrations that matters where in this list it's placed?

camsaul replied:
sorry, you're right
carry on 👍

camsaul commented on src/metabase/driver/druid.clj:
```diff
@@ -11,7 +11,8 @@
             [metabase.sync-database.analyze :as analyze]
-            [metabase.util.ssh :as ssh]))
+            [metabase.util.ssh :as ssh]
+            [metabase.sync-database.cached-values :as cached-values]))
```
sorting

camsaul commented on src/metabase/driver/generic_sql.clj:
```diff
@@ -18,7 +18,8 @@
              [honeysql-extensions :as hx]
-             [ssh :as ssh]])
+             [ssh :as ssh]]
+            [metabase.sync-database.cached-values :as cached-values])
```
sorting

camsaul commented on src/metabase/driver/generic_sql.clj:
```diff
@@ -454,7 +448,8 @@
    :field->alias         (u/drop-first-arg name)
-   :field-percent-urls   fast-field-percent-urls
+;   :field-percent-urls   fast-field-percent-urls
+;   :field-avg-length     (u/drop-first-arg field-avg-length)
```
why is this all getting commented out? Do we not need it anymore? (If so, why not just delete the lines altogether?)

camsaul commented on src/metabase/driver/mongo.clj:
```diff
@@ -21,7 +21,9 @@
-            [toucan.db :as db])
+            [toucan.db :as db]
+            [metabase.sync-database.classify :as classify]
+            [metabase.sync-database.cached-values :as cached-values])
```
sorting

camsaul commented:
@thearthur by default CI doesn't run tests against all the drivers unless you include `[ci drivers]` in the commit message so I'd recommend doing that at some point to make sure things work for every driver. Otherwise it's easy to get into a trap where you got everything working for H2 but broke everything else and have to spend forever fixing things at the end

camsaul commented on src/metabase/sync_database.clj:
```diff
@@ -12,60 +12,13 @@
-(declare sync-database-with-tracking!
-         sync-table-with-tracking!)
+            [toucan.db :as db]
+            [metabase.sync-database.classify :as classify]))
```
sorting

camsaul commented:
also looks like you're getting CI failures from the linter for trailing whitespace. Don't forget to 

```emacs-lisp
(add-hook 'before-save-hook #'delete-trailing-whitespace)
```

to your `init.el`

camsaul commented on src/metabase/sync_database/analyze.clj:
```diff
@@ -157,86 +59,99 @@
+                      (:field_id fingerprint) (:name fingerprint) (keys fingerprint)))
+         (or (db/update! FieldFingerprint {:where [:= :field_id (:field_id fingerprint)]
+                                           :set fingerprint})
+             (db/insert! FieldFingerprint fingerprint)))))
```
this is annoying but you could potentially hit a race condition here if two threads try to save a new fingerprint for the same field ID at the same. They could both try to hit `insert!` at the same time and one will fail because it's now a duplicate entry.

camsaul commented on src/metabase/sync_database/analyze.clj:
```diff
@@ -157,86 +59,99 @@
+  [fingerprint]
+  (or (db/update! TableFingerprint {:where [:= :table_id (:table_id fingerprint)]
+                                    :set fingerprint})
+      (db/insert! TableFingerprint fingerprint)))
```
also a potential race condition

camsaul commented on src/metabase/sync_database/cached_values.clj:
```diff
@@ -0,0 +1,136 @@
+            [toucan.db :as db]
+            [metabase.db.metadata-queries :as queries]
+            [metabase.sync-database.classify :as classify]
+            [metabase.sync-database.infer-special-type :as infer-special-type]))
```
ns sorting

camsaul commented on src/metabase/sync_database/cached_values.clj:
```diff
@@ -0,0 +1,136 @@
+             :or   {calculate-row-count?  true}}]
+  (fn [driver table new-field-ids]
+    {:fields    (for [id new-field-ids]
+                  (extract-field-values (field/Field id) {:id id}))}))
```
So it looks like `calculate-row-count?` isn't even being used for anything. What's the reason for having it?

Also why do we need this `make-field-extractor` function at all? It seems like every time it more or less always produces the same function. Seems to me like it could be removed altogether and its behavior could just be rolled into `generic-analyze-table`

salsakran replied:
not sure, but if we can capture row count for tables, we should. Not sure where that is best done, but it's tricky because for some DBs its almost free, and for others it's expensive.

camsaul commented on src/metabase/sync_database/sync.clj:
```diff
@@ -12,7 +12,11 @@
+            [metabase.sync-database :as sync-database]
+            [metabase.sync-database.cached-values :as cached-values]
+            [metabase.sync-database.analyze :as analyze]
+            [metabase.sync-database.classify :as classify])
```
ns sorting

camsaul commented on src/metabase/sync_database/sync.clj:
```diff
@@ -275,3 +279,25 @@
     ;; HACK! we can't sync the _metabase_metadata table until all the "Raw" Tables/Columns are backed
     (maybe-sync-metabase-metadata-table! database raw-tables)))
+
+(defn future-sync-and-analyze-table
```
I think this function should probably be called something like `sync-and-analyze-table-async!`. 

1.  It has side-effects -- calling it twice is not the same as calling it once. (Calling it twice would mean a table would get synced twice.) https://github.com/bbatsov/clojure-style-guide#changing-state-fns-with-exclamation-mark
2.  The fact that it uses a `future` is an implementation detail of it being asynchronous. Saying `async` instead of `future` hides those implementation details and leaves open the door we might implement it in some other way in the future without having to change the name of it. Also I think people less familiar with Clojure would not necessarily know what `future-` meant right away but `-async` should be clear to just about everyone

camsaul commented on test/metabase/api/public_test.clj:
```diff
@@ -12,13 +12,16 @@
             [toucan.db :as db]
-            [toucan.util.test :as tt])
+            [toucan.util.test :as tt]
+            [metabase.sync-database :as sync-database])
```
ns sorting

camsaul commented on test/metabase/api/table_test.clj:
```diff
@@ -20,7 +20,10 @@
+            [toucan.util.test :as tt]
+            [metabase.sync-database.cached-values :as cached-values]
+            [metabase.sync-database.classify :as classify]
+            [metabase.sync-database.analyze :as analyze]))
```
ns sorting

camsaul commented on test/metabase/driver/generic_sql_test.clj:
```diff
@@ -9,7 +9,8 @@
             [metabase.test.data.datasets :as datasets]
-            [toucan.db :as db])
+            [toucan.db :as db]
+            [metabase.sync-database.cached-values :as cached-values])
```
ns sorting

camsaul commented on test/metabase/models/field_values_test.clj:
```diff
@@ -26,7 +26,7 @@
-(expect false (field-should-have-field-values? {:special_type    nil
+(expect true (field-should-have-field-values? {:special_type    nil
                                                 :visibility_type :normal
                                                 :base_type       :type/Text}))
```
why did this change from `false` to `true`? (Not saying it's wrong, just trying to understand the changes that are being made in this PR)

camsaul commented on test/metabase/sync_database_test.clj:
```diff
@@ -19,7 +19,11 @@
+            [toucan.util.test :as tt]
+            [metabase.sync-database.cached-values :as cached-values]
+            [clojure.tools.logging :as log]
+            [metabase.sync-database.analyze :as analyze]
```
ns sorting

camsaul commented on test/metabase/sync_database_test.clj:
```diff
@@ -142,10 +147,17 @@
     (sync-database! db)
+    (cached-values/cache-field-values-for-database! db)
+    (analyze/analyze-database db)
+    (classify/classify-database! db)
```
should `analyze-database` end with a `!`?

thearthur replied:
yep, fixed!

camsaul commented on test/metabase/sync_database_test.clj:
```diff
@@ -292,43 +309,9 @@
+;;; -------------------- Make sure that if a Field's cardinality passes `metabase.sync-database.classify/low-cardinality-threshold` (currently 300) (#3215) --------------------
 (defn- insert-range-sql [rang]
-  (str "INSERT INTO blueberries_consumed (num) VALUES "
+  (str "INSERT INTO blueberries_consumed_rating (num) VALUES "
```
`blueberries_consumed_rating`?

thearthur replied:
I'm trying to figure out where this came from. slightly disturbing

camsaul replied:
Maybe you changed it to `_rating` so it would get marked as a category?

salsakran replied:
this seems like it's incorrect

camsaul commented on test/metabase/test/data/postgres.clj:
```diff
@@ -21,13 +21,21 @@
          (when (= context :db)
-           {:db database-name})))
+           {:db database-name})
+         (when-let [password (env :mb-test-db-pass)]
```
I think when we talked about this I suggested names like `MB_POSTGRES_TEST_DB_PASS` or the like. Because otherwise it's not obvious something like `MB_TEST_DB_PASS` is Postgres-specific. I'm in favor of adding the new env vars for configuration purposes but the need to have something in them marking them as Postgres-specific in case we want to do the same thing for MySQL

thearthur replied:
good point, i'll change the name.

camsaul commented:
My big question is... do we have any new tests for the sync scheduling behavior? e.g. something like 

1.  Add a new database with some specific sync schedule (maybe a few seconds or something)
2.  Sleep for `n` number of seconds
3.  See if it did indeed get synced as promised

camsaul commented on src/metabase/api/table.clj:
```diff
@@ -10,11 +10,16 @@
-             [hydrate :refer [hydrate]]]))
+             [hydrate :refer [hydrate]]]
+            [metabase.sync-database.classify :as classify]
+            [metabase.driver :as driver]))
```
ns sorting 🙀

camsaul commented on src/metabase/api/table.clj:
```diff
@@ -66,14 +71,18 @@
                      :description             description))
     (api/check-500 (db/update! Table id, :visibility_type visibility_type))
     (let [updated-table (Table id)
+          driver (->> id table/table-id->database-id driver/database-id->driver) ;; is this the easy way to do this?
```
You can just do `(driver/database-id->driver (:db_id updated-table))`. Otherwise you're doing an extra DB call here

camsaul commented on src/metabase/driver/bigquery.clj:
```diff
@@ -23,7 +23,8 @@
             [metabase.util.honeysql-extensions :as hx]
-            [toucan.db :as db])
+            [toucan.db :as db]
+            [metabase.sync-database.cached-values :as cached-values])
```
ns sorting

thearthur replied:
✔️

camsaul commented on src/metabase/driver/generic_sql.clj:
```diff
@@ -312,7 +308,7 @@
+#_(defn fast-field-percent-urls
   "Fast, default implementation of `field-percent-urls` that avoids a full table scan."
   [driver field]
   (let [table       (field/table field)
```
all these things that are commented out, is that because they're no longer used? If so, we should probably just delete the lines instead

thearthur replied:
code polution, cleaned it up

camsaul commented on src/metabase/driver/mongo.clj:
```diff
@@ -148,11 +150,12 @@
    :fields    (for [{:keys [id] :as field} (table/fields table)
-                    :when (analyze/test-for-cardinality? field (contains? new-field-ids (:id field)))]
-                (analyze/test:cardinality-and-extract-field-values field {:id id}))})
+                    #_:when #_(classify/test-for-cardinality? field (contains? new-field-ids (:id field)))]
```
can we remove all the commented out stuff, or is there a reason to keep it around

thearthur replied:
nope, it's left from my iterative refactoring flow,  removed ✔️

camsaul commented on src/metabase/driver/presto.clj:
```diff
@@ -19,7 +19,8 @@
              [honeysql-extensions :as hx]
-             [ssh :as ssh]])
+             [ssh :as ssh]]
+            [metabase.sync-database.cached-values :as cached-values])
```
ns sorting

thearthur replied:
✔️

camsaul commented on src/metabase/driver/presto.clj:
```diff
@@ -124,7 +125,7 @@
-  ((analyze/make-analyze-table driver
+  ((cached-values/make-field-extractor driver
      :field-avg-length-fn   field-avg-length
      :field-percent-urls-fn field-percent-urls) driver table new-table-ids))
```
Wait, `make-field-extractor` doesn't take `field-avg-length-fn` and `field-percent-urls-fn` anymore, right? This looks wrong

thearthur replied:
yep that's doing nothing there, needs fixing

thearthur replied:
✔️

camsaul commented on src/metabase/events/sync_database.clj:
```diff
@@ -15,31 +14,47 @@
 
-;;; ## ---------------------------------------- EVENT PROCESSING ----------------------------------------
+(def ^:private sync-delete-channel
+  "Channel for receiving event notifications we want to subscribe to for database sync events."
```
This docstring looks like copypasta https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#make-sure-to-update-comments

thearthur replied:
👍fixed

kdoh committed:
initial scheduling ui

thearthur committed:
cleanup from review

thearthur committed:
Merge remote-tracking branch 'origin/sync-analyze-cache-split' into sync-analyze-cache-split

thearthur committed:
style cleanup

sort test data tables

thearthur committed:
rename analyze/analyze-database

add a ! to the end to match similar functions

thearthur committed:
finish removing unused driver references.

thearthur committed:
calculate % email and json in fingerprints

thearthur committed:
Merge branch 'master' into sync-analyze-cache-split

attekei commented:
Working now on wiring up the scheduling/sync admin section UI! Kyle has already done the UI side, so I’m working on adding the needed links between the UI and backend. This is the latest design:
![screen shot 2017-06-28 at 14 50 36](https://user-images.githubusercontent.com/6034928/27663602-00f69e9e-5c18-11e7-88b8-90353c76bd07.png)

It’s already pretty clear to me how to:
- change the database sync interval ( => I change the `sync_schedule` value accordingly)
- how to implement the “Often” choice in field in field fingerprinting ( => I change `cache_field_values_schedule` value based on granularity + date time combination)

(I will probably write / find from Github some lightweight wrapper around manipulating crontab values in react components / redux actions)

What is still unclear:
- What is the relation of `analyze_schedule` and `classify_schedule` schedules to the admin UI? Should updating the values of Database Sync / Field Fingerprinting sections update also those?
- How the choice between “Often”, “Rarely” and “Never” settings should be saved? Should we create some enum field for those (and the field fingerprinting cron setting would only be taken into account if the enum value is `often` or similar)?
- How “Discard saved field values” action should be implemented and is there already an API endpoint for doing that?
- Is there yet a matching API endpoint for “Re-scan field values now” action?
- How “Discard saved field values” action should be implemented and is there already an API endpoint for that?

salsakran commented:
@attekei feel free to stub out a backend api endpoint until @thearthur, @camsaul or myself can get around to it.

attekei committed:
Add mock endpoints and let PUT /database/id accept a trigger type param

attekei committed:
Boostrapping frontend testing

attekei committed:
Add failing tests for database admin UI

thearthur committed:
formating

thearthur committed:
Merge remote-tracking branch 'origin/sync-analyze-cache-split' into sync-analyze-cache-split

attekei commented:
Frontend has been wired up now! See the end of metabase.api.database for the endpoints I still need for triggering manual schema update, manual field value scan and the deletion of saved field values.

I used the same format as pulse schedule settings use for `sync_schedule` and `cache_field_values_schedule` values, which is like this:
```
{ schedule_day: "mon", schedule_frame: "null", schedule_hour: 4, schedule_type: "daily" }
```
The backend needs to be modified to support this format in API endpoints instead of cron strings (i.e. the sync schedule parsing logic should maybe be similar as what we have for pulses).

attekei committed:
Somewhat working frontend for db sync settings

camsaul committed:
Merge branch 'master' into sync-analyze-cache-split

camsaul commented on test/metabase/test/data/postgres.clj:
```diff
@@ -21,13 +21,21 @@
-          :port     5432
+  (merge {:host     (or (env :mb-test-db-host)
+                        "localhost")
+          :port     (or (env :mb-test-db-port)
```
this needs to be changed to something like `:mb-pg-test-db-port`

camsaul committed:
Fix namespace :require sorting :unamused:

camsaul committed:
Indentation fixes :wrench: :unamused:

camsaul reviewed (changes requested):
Need to fill placeholder endpoints

salsakran commented on src/metabase/api/database.clj:
```diff
@@ -365,19 +373,43 @@
+(api/defendpoint POST "/:id/discard_values"
+  "Discards all saved field values for this `Database`."
+  [id]
+  {:status :ok})
```
yup, that's what I was asking you to do =)

@attekei wrote some stubs to get roundtrips on the frontend working

thearthur committed:
Revert "style cleanup"

This reverts commit ab4e5becdad22b66ab9ed6a8429376ef736cafaa.

camsaul commented:
looks like we need to make sure we're binding `*disable-qp-logging*` to `true` when running MBQL queries during analysis. Otherwise the logs are filled with QP related stuff which makes debugging sync hard

attekei committed:
Add some action points as a result of chatting with Kyle

camsaul committed:
schedule-map <-> cron-string util fns

camsaul committed:
Merge branch 'sync-analyze-cache-split' of github.com:metabase/metabase into sync-analyze-cache-split

attekei committed:
Only produce valid and unambiguous outputs in SchedulePicker

attekei committed:
Fix accidental "null"s and replace with plain nulls

attekei committed:
Fix typo, use midnight as default hour

camsaul committed:
Clarify purpose of var with better name

camsaul committed:
Fix lint issue :shirt:

camsaul committed:
Merge branch 'master' into sync-analyze-cache-split

camsaul committed:
Fix lint error (trailing whitespace)

salsakran commented on test/metabase/driver/presto_test.clj:
```diff
@@ -123,12 +123,6 @@
-  0.5
-  (data/dataset half-valid-urls
-    (sql/field-percent-urls (PrestoDriver.) (db/select-one 'Field :id (data/id :urls :url)))))
-
```
we are switching from checking for a specific threshold and marking something as a url to keeping track of the fraction of entries that are urls, putting that in a fingerprint and using that to classify things later.

camsaul replied:
thanks 👍

salsakran commented on src/metabase/sync_database/analyze.clj:
```diff
@@ -251,6 +169,7 @@
         finished-tables-count (atom 0)]
     (doseq [{table-name :name, :as table} tables]
       (try
+        #_(cached-values/cache-table-data-shape! driver table)
```
not sure. The expected logic is 
1. field values are cached on a schedule
2. When first setting up a DB we should do all three actions (sync, cache-values, analyze->classify)
3. Any time something is analyzed it should be fed to the classifier

In the future we'll tweak all of these.

camsaul committed:
Merge branch 'master' into sync-analyze-cache-split

camsaul committed:
Better fn name

camsaul committed:
Test fix :wrench:

camsaul committed:
Merge branch 'master' into sync-analyze-cache-split

camsaul commented on src/metabase/task/sync.clj:
```diff
@@ -0,0 +1,117 @@
+(def ^:private ^:const classify-databases-trigger-key "metabase.task.%s-databases.trigger-%s")
+
+(defonce ^:private classify-databases-job (atom nil))
+(defonce ^:private classify-databases-trigger (atom nil))
```
This PR is becoming increasingly scary to me as I realize there are all sorts of things like this var here that are defined but never used *anywhere*. It's hard to tell whether it's leftover code that should have been removed or whether some of the other code is written incorrectly because and should be using this but isn't

I'm worried that if we merge this for 25 it will break something and I'll be lost as to how to fix it because I can't understand what's supposed to be happening and what's not working right

camsaul commented on src/metabase/sync_database/analyze.clj:
```diff
@@ -7,236 +7,154 @@
+                                  more)))))
+
+(defn field-avg-length
+  "Default implementation of optional driver fn `field-avg-length` that calculates the average length in Clojure-land via `field-values-lazy-seq`."
```
...functions whose purpose has completely changed but without corresponding updates to the docstrings. Makes it hard to figure out how this is all supposed to work 😓

camsaul commented on src/metabase/driver/presto.clj:
```diff
@@ -350,7 +348,6 @@
           :current-datetime-fn       (constantly :%now)
           :date                      (u/drop-first-arg date)
           :excluded-schemas          (constantly #{"information_schema"})
-          :field-percent-urls        (u/drop-first-arg field-percent-urls)
```
... this is just crazy. So it seems like `field-percent-urls` is being taking out of the `IDriver` interface (which is fine) but the corresponding implementations are just left there, doing nothing. Dead code *all over the place* like that makes things extremely confusing. Plus I've had to clean up so much of this code I'm worried I may have inadvertently broke something

camsaul commented on src/metabase/sync_database/cached_values.clj:
```diff
@@ -0,0 +1,136 @@
+(defn generic-analyze-table
+  "An implementation of `analyze-table` using the defaults (`default-field-avg-length` and `field-percent-urls`)."
+  [driver table new-field-ids]
+  ((make-field-extractor driver) driver table new-field-ids))
```
here's another function whose purpose completely changed but with a name that no longer makes any sense (?) and a docstring that makes even less sense

salsakran commented:
@camsaul might make sense to start from test coverage and work backwards? Eg extend the current tests so that we have confidence it's doing (or not doing) what it should.

camsaul commented on src/metabase/driver/generic_sql.clj:
```diff
@@ -430,9 +376,7 @@
-  ((analyze/make-analyze-table driver
-     :field-avg-length-fn   (partial field-avg-length driver)
-     :field-percent-urls-fn (partial field-percent-urls driver))
+  ((cached-values/make-field-extractor driver)
```
other insanity... calling a function that returns a function that is then used to get the field values. Instead of just making that function top-level and calling it directly. Luckily, this code is only copy-pasted 4 times in the code base, and only one of them is passing the `:calculate-row-count? false` options that aren't used *anywhere*.

camsaul commented on src/metabase/sync_database/cached_values.clj:
```diff
@@ -0,0 +1,136 @@
+  {:style/indent 1}
+  [driver & {:keys [calculate-row-count?]
+             :or   {calculate-row-count?  true}}]
+  (fn [driver table new-field-ids]
```
why are we passing around unused arguments like `table` here. just. why.

camsaul replied:
oh yeah. `driver` isn't used either! 😂 😂 😂 

I think these fn args must have just been left in place because the old function that sort of did something similar needed them

camsaul commented on src/metabase/sync_database/cached_values.clj:
```diff
@@ -0,0 +1,136 @@
+(defn cache-data-shape-for-tables!
+  "Perform in-depth analysis on the data shape for all `Tables` in a given DATABASE.
+   This is dependent on what each database driver supports, but includes things like cardinality testing and table row counting.
+   The bulk of the work is done by the `(analyze-table ...)` function on the IDriver protocol."
```
Copy-pasta-ing this out-of-date docstring from the old sync code to here and to the `analyze-data-shape-for-tables!` function doesn't do much to make the new sync code easier to work with. 😓

camsaul replied:
i'm going around trying to clean things up and remove functions that aren't needed and then I see them mentioned in the docstrings of other functions that don't use them. And then I see that same docstring copy-pasta-ed everywhere. 💔

camsaul replied:
omg actually the whole function is basically cut and pasted. I am officially confused.

Two functions that both "analyze data in [a] database"

camsaul replied:
nvm. The logging messages were just not updated when the functions was copy-pastaed and tweaked

camsaul committed:
Code cleanup (part 5 of many more to go)

camsaul commented on src/metabase/cache_database.clj:
```diff
@@ -0,0 +1,75 @@
+  (let [start-time (System/nanoTime)
+        tracking-hash (str (java.util.UUID/randomUUID))]
+    (log/info (u/format-color 'magenta "Syncing %s database '%s'..." (name driver) (:name database)))
+    (events/publish-event! :database-analysis-begin {:database_id (:id database) :custom_id tracking-hash})
```
so for caching a "database data shape" we publish an event called `database-analysis-begin`? Does that make any sense?

salsakran replied:
This this is a case of re-using old names

camsaul committed:
Code cleanup (part 6 of ?)

camsaul commented on test/metabase/models/field_test.clj:
```diff
@@ -4,82 +4,3 @@
-(expect :type/Category (infer-field-special-type "rating"  :type/Integer))
-(expect nil            (infer-field-special-type "rating"  :type/Boolean))
-(expect :type/Country  (infer-field-special-type "country" :type/Text))
-(expect nil            (infer-field-special-type "country" :type/Integer))
```
deleting several dozen tests = 😱

camsaul committed:
Cleanup part 7

camsaul committed:
Cleanup part 8

camsaul committed:
Code cleanup 9

camsaul committed:
Code cleanup 10

camsaul commented:
my new favorite thing in this PR: an entire namespace is copied over (~300 LOC) and the only difference is one version has an extra couple of functions

camsaul commented:
```
GeoTrip:metabase camsaul$ diff sync_database/sync_schema.clj  sync_database/sync.clj 
1c1
< (ns metabase.sync-database.sync-schema
---
> (ns metabase.sync-database.sync
8a9
>              [sync-database :as sync-database]
14a16,19
>             [metabase.sync-database
>              [analyze :as analyze]
>              [cached-values :as cached-values]
>              [classify :as classify]]
277a283,304
> 
> (defn sync-and-analyze-table-async!
>   "run the entire sync process for a table in a future"
>   [table]
>   (try
>     (future (sync-database/sync-table! table)
>             (cached-values/cache-field-values-for-table! table)
>             (analyze/analyze-table-data-shape! table)
>             (classify/classify-table! table))
>     (catch Throwable e
>         (log/error (format "Error syncing Table: %d" (:id table))))))
> 
> (defn future-sync-and-analyze-database
>   "run the entire sync process for a database in a future"
>   [database]
>   (try
>     (future (sync-database/sync-database! database)
>             (cached-values/cache-field-values-for-database! database)
>             (analyze/analyze-database! database)
>             (classify/classify-database! database))
>     (catch Throwable e
>         (log/error (format "Error syncing Database: %d" (:id database))))))

```

camsaul committed:
Code cleanup 10

camsaul commented:
other fun news

the tests have never passed on this branch since the inception of the PR

and I'm having a lot of trouble figuring out why since it just hangs at random spots when you run the tests

camsaul committed:
Code cleanup 11

camsaul committed:
Debug logging (etc)

camsaul committed:
Code cleanup 12

camsaul commented:
This PR is completely out-of-hand with 4600 LOC changes. It makes tracking something like mystery test hanging next to impossible. So I'm splitting this PR up into several pieces in a last-ditch effort to salvage this PR. So far broken into #5463 and #5464, but more splits might be done in the future
